File: python/tests/test-compoundparser.py
import os
if "EXTRA_DLL_SEARCH_PATHS" in os.environ and hasattr(os, "add_dll_directory"):
    for path in os.environ["EXTRA_DLL_SEARCH_PATHS"].split(os.pathsep):
        os.add_dll_directory(path)
import unittest
import xraylib


class TestCompoundParser(unittest.TestCase):
    def test_good_compounds(self):
        self.assertIsInstance(xraylib.CompoundParser("C19H29COOH"), dict)
        self.assertIsInstance(xraylib.CompoundParser("C12H10"), dict)
        self.assertIsInstance(xraylib.CompoundParser("C12H6O2"), dict)
        self.assertIsInstance(xraylib.CompoundParser("C6H5Br"), dict)
        self.assertIsInstance(xraylib.CompoundParser("C3H4OH(COOH)3"), dict)
        self.assertIsInstance(xraylib.CompoundParser("HOCH2CH2OH"), dict)
        self.assertIsInstance(xraylib.CompoundParser("C5H11NO2"), dict)
        self.assertIsInstance(xraylib.CompoundParser("CH3CH(CH3)CH3"), dict)
        self.assertIsInstance(xraylib.CompoundParser("NH2CH(C4H5N2)COOH"), dict)
        self.assertIsInstance(xraylib.CompoundParser("H2O"), dict)
        self.assertIsInstance(xraylib.CompoundParser("Ca5(PO4)3F"), dict)
        self.assertIsInstance(xraylib.CompoundParser("Ca5(PO4)3OH"), dict)
        self.assertIsInstance(xraylib.CompoundParser("Ca5.522(PO4.48)3OH"), dict)
        self.assertIsInstance(xraylib.CompoundParser("Ca5.522(PO.448)3OH"), dict)

    def test_bad_compounds(self):
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("CuI2ww")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("0C")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("2O")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("13Li")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("2(NO3)")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("H(2)")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Ba(12)")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Cr(5)3")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Pb(13)2")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au(22)11")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au11(H3PO4)2)")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au11(H3PO4))2")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au(11(H3PO4))2")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Ca5.522(PO.44.8)3OH")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Ba[12]")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Auu1")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("AuL1")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser(None)
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("  ")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("\t")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("\n")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au L1")
        with self.assertRaises(ValueError):
            xraylib.CompoundParser("Au\tFe")
        with self.assertRaises(TypeError):
            xraylib.CompoundParser(26)

    def test_H2SO4(self):
        cd = xraylib.CompoundParser('H2SO4')
        self.assertEqual(cd['nElements'], 3)
        self.assertEqual(cd['molarMass'], 98.09)
        self.assertEqual(cd['nAtomsAll'], 7.0)
        self.assertEqual(cd['Elements'], (1, 8, 16))
        self.assertAlmostEqual(cd['massFractions'], (0.02059333265368539, 0.6524620246712203, 0.32694464267509427))
        self.assertAlmostEqual(cd['nAtoms'], (2.0, 4.0, 1.0))

class TestSymbolToAtomicNumber(unittest.TestCase):
    def test_Fe(self):
        self.assertEqual(xraylib.SymbolToAtomicNumber('Fe'), 26)

    def test_bad_symbol(self):
        with self.assertRaises(ValueError):
            xraylib.SymbolToAtomicNumber('Uu')

    def test_bad_type(self):
        with self.assertRaises(TypeError):
            xraylib.SymbolToAtomicNumber(26)
        with self.assertRaises(ValueError):
            xraylib.SymbolToAtomicNumber(None)

class TestAtomicNumberToSymbol(unittest.TestCase):
    def test_Fe(self):
        self.assertEqual(xraylib.AtomicNumberToSymbol(26), 'Fe')

    def test_bad_symbol(self):
        with self.assertRaises(ValueError):
            xraylib.AtomicNumberToSymbol(-2)
        with self.assertRaises(ValueError):
            xraylib.AtomicNumberToSymbol(108)

    def test_bad_type(self):
        with self.assertRaises(TypeError):
            xraylib.AtomicNumberToSymbol("26")
        with self.assertRaises(TypeError):
            xraylib.AtomicNumberToSymbol("Fe")
        with self.assertRaises(TypeError):
            xraylib.AtomicNumberToSymbol(None)

class TestCrossValidation(unittest.TestCase):
    def test(self):
        for Z in range(1, 108):
            symbol = xraylib.AtomicNumberToSymbol(Z)
            self.assertEqual(xraylib.SymbolToAtomicNumber(symbol), Z)

if __name__ == '__main__':
    unittest.main(verbosity=2)

File: lua/tests/test-compoundparser.lua
require("xraylib")
luaunit = require("luaunit")

TestCompoundParser = {}
	function TestCompoundParser:test_good_compounds()
		luaunit.assertIsTable(xraylib.CompoundParser("C19H29COOH"))
		luaunit.assertIsTable(xraylib.CompoundParser("C12H10"))
		luaunit.assertIsTable(xraylib.CompoundParser("C12H6O2"))
		luaunit.assertIsTable(xraylib.CompoundParser("C6H5Br"))
		luaunit.assertIsTable(xraylib.CompoundParser("C3H4OH(COOH)3"))
		luaunit.assertIsTable(xraylib.CompoundParser("HOCH2CH2OH"))
		luaunit.assertIsTable(xraylib.CompoundParser("C5H11NO2"))
		luaunit.assertIsTable(xraylib.CompoundParser("CH3CH(CH3)CH3"))
		luaunit.assertIsTable(xraylib.CompoundParser("NH2CH(C4H5N2)COOH"))
		luaunit.assertIsTable(xraylib.CompoundParser("H2O"))
		luaunit.assertIsTable(xraylib.CompoundParser("Ca5(PO4)3F"))
		luaunit.assertIsTable(xraylib.CompoundParser("Ca5(PO4)3OH"))
		luaunit.assertIsTable(xraylib.CompoundParser("Ca5.522(PO4.48)3OH"))
		luaunit.assertIsTable(xraylib.CompoundParser("Ca5.522(PO.448)3OH"))
	end

	function TestCompoundParser:test_bad_compounds()
		luaunit.assertError(xraylib.CompoundParser, "CuI2ww")
		luaunit.assertError(xraylib.CompoundParser, "0C")
		luaunit.assertError(xraylib.CompoundParser, "2O")
		luaunit.assertError(xraylib.CompoundParser, "13Li")
		luaunit.assertError(xraylib.CompoundParser, "2(NO3)")
		luaunit.assertError(xraylib.CompoundParser, "H(2)")
		luaunit.assertError(xraylib.CompoundParser, "Ba(12)")
		luaunit.assertError(xraylib.CompoundParser, "Cr(5)3")
		luaunit.assertError(xraylib.CompoundParser, "Pb(13)2")
		luaunit.assertError(xraylib.CompoundParser, "Au(22)11")
		luaunit.assertError(xraylib.CompoundParser, "Au11(H3PO4)2)")
		luaunit.assertError(xraylib.CompoundParser, "Au11(H3PO4))2")
		luaunit.assertError(xraylib.CompoundParser, "Au(11(H3PO4))2")
		luaunit.assertError(xraylib.CompoundParser, "Ca5.522(PO.44.8)3OH")
		luaunit.assertError(xraylib.CompoundParser, "Ba[12]")
		luaunit.assertError(xraylib.CompoundParser, "Auu1")
		luaunit.assertError(xraylib.CompoundParser, "AuL1")
		luaunit.assertError(xraylib.CompoundParser, nil)
		luaunit.assertError(xraylib.CompoundParser, "  ")
		luaunit.assertError(xraylib.CompoundParser, "\t")
		luaunit.assertError(xraylib.CompoundParser, "\n")
		luaunit.assertError(xraylib.CompoundParser, "Au L1")
		luaunit.assertError(xraylib.CompoundParser, "Au\tFe")
		luaunit.assertError(xraylib.CompoundParser, 26)
	end

	function TestCompoundParser:test_H2SO4()
		cd = xraylib.CompoundParser("H2SO4")
		luaunit.assertEquals(cd['nElements'], 3)
		luaunit.assertAlmostEquals(cd['molarMass'], 98.09)
		luaunit.assertAlmostEquals(cd['nAtomsAll'], 7.0)
		luaunit.assertEquals(cd['Elements'], {1, 8, 16})
		luaunit.assertAlmostEquals(cd['massFractions'][1], 0.02059333265368539, 1E-6)
		luaunit.assertAlmostEquals(cd['massFractions'][2], 0.6524620246712203, 1E-6)
		luaunit.assertAlmostEquals(cd['massFractions'][3], 0.32694464267509427, 1E-6)
		luaunit.assertEquals(cd['nAtoms'], {2.0, 4.0, 1.0})
	end

TestSymbolToAtomicNumber = {}
	function TestSymbolToAtomicNumber:test_Fe()
		luaunit.assertEquals(xraylib.SymbolToAtomicNumber('Fe'), 26)
	end

	function TestSymbolToAtomicNumber:test_bad_symbol()
		luaunit.assertError(xraylib.SymbolToAtomicNumber, 'Uu')
	end

	function TestSymbolToAtomicNumber:test_bad_type()
		luaunit.assertError(xraylib.SymbolToAtomicNumber, 26)
	end

TestAtomicNumberToSymbol = {}
	function TestAtomicNumberToSymbol:test_Fe()
		luaunit.assertEquals(xraylib.AtomicNumberToSymbol(26), 'Fe')
		luaunit.assertEquals(xraylib.AtomicNumberToSymbol("26"), 'Fe')
	end

	function TestAtomicNumberToSymbol:test_bad_symbol()
		luaunit.assertError(xraylib.AtomicNumberToSymbol, -2)
		luaunit.assertError(xraylib.AtomicNumberToSymbol, 108)
	end

	function TestAtomicNumberToSymbol:test_bad_type()
		luaunit.assertError(xraylib.AtomicNumberToSymbol, "Fe")
		luaunit.assertError(xraylib.AtomicNumberToSymbol, nil)
	end

TestCrossValidation = {}
	function TestCrossValidation:test()
		for Z = 1, 107 do
			symbol = xraylib.AtomicNumberToSymbol(Z)
			luaunit.assertEquals(xraylib.SymbolToAtomicNumber(symbol), Z)
		end
	end

os.exit(luaunit.LuaUnit.run('-v', '-o', 'tap'))
